### DISCUSSION TITLE
导入txt词典以后，出现大量联想造词

### DISCUSSION THREAD

**Discussion by ghsgz** [fcitx5-android/fcitx5-android#260] [Q&A | 问答]
> 我没有在issue里找到相似的问题。

经过简单测试，这个问题可以稳定复现，取消选择导入的词典以后，问题就消失了。

## 问题描述

以下是截图说明（用于测试的词语是：艾儿天空、发财）。

![Screenshot_20230507_153603_com microsoft emmx](https://user-images.githubusercontent.com/34056529/236667668-a4fc5a60-aa8b-4704-8440-d93006f43d89.png)

![Screenshot_20230507_153534_com microsoft emmx](https://user-images.githubusercontent.com/34056529/236667655-9621ecf1-c06c-48c2-9ff9-eb54535e16ac.png)

## 上下文信息

版本信息：0.0.5-0-g679ab7db-release

输入法方案：小鹤双拼

导入的词库的内容片段：

```
……
艾儿天空 ai'er'tian'kong 7
……
发财 fa'cai 2
```

**Comment by rocka** [comment]
导入以下词库：

```
艾儿天空 ai'er'tian'kong 7
发财 fa'cai 2
```

在我的设备上没有复现

![Screenshot_20230507-173105](https://user-images.githubusercontent.com/13914967/236669556-b1239e6b-7202-4860-a914-a1b50bc08455.png)

建议更新到最新的 CI build 版本再试一下 https://jenkins.fcitx-im.org/job/android/job/fcitx5-android/

如果还是能复现，请提供完整的词库，而不是只有两个词的片段

**Reply by ghsgz** [reply]
覆盖安装最新版本以后，重新测试了一下，情况如下：

- 如果词库很小（只有一百行），候选词表现正常；
- 如果是原词库（23234个词），问题重新出现。

词库的基本信息：通过深蓝词库转换程序，将微软拼音自学习词库、搜狗备份词库、手心输入法分别转化为手心txt文件，然后合并词库，输出为符合fcitx5要求的txt格式：词语 拼音编码 词频）。

这个词库可以正常导入fcitx-for-android，之前转换与合并词库时，因为换行符\r\n等问题，导致导入时出错，已经解决导入时的问题了。

所以，这个问题是和我的词库有关？还是和词库大小有关？我不太了解fcitx5，不过我知道微软自学习词库有上限，是两万条，超过会抹除已有的词语。

**Reply by rocka** [reply]
理论上来说词库大小是没有上限的。建议检查一下“联想”出来的那些“词”是否存在于导入的词库中。

**Reply by ghsgz** [reply]
我试过了，导入的词库里没有这些预期之外的词语，应该是由fcitx5的引擎或扩展联想组词生成的。
词库在桌面端的微软拼音输入法、手心输入法里都表现正常。
通过深蓝词库重新转换成手心输入法，导入移动端的手心输入法，手心输入法的表现符合预期。

也许和“附加组件——配置附加组件——拼写——后端”有关？与“预测”相关的功能，我也是关闭的。

**Reply by rocka** [reply]
“拼写”是指英文单词拼写检查。“预测”是指选词上屏之后出现的联想，而不是输入中出现的候选词。

**Comment by llpub** [comment]
全拼方案导入词库后也有同样问题。
但是测试了一下，修改导入词库的词频，最大不超过2，这种莫明其妙组合词的情况就能极大缓解

**Reply by Marukon** [reply]
> 全拼方案导入词库后也有同样问题。 但是测试了一下，修改导入词库的词频，最大不超过2，这种莫明其妙组合词的情况就能极大缓解

我也发现了，如果词库的词频过大（几万甚至几十万），就会出现各种奇怪的组合

**Comment by lunzima** [comment]
libime的词频是log10对数。
见https://github.com/fcitx/fcitx5-chinese-addons/issues/219

**Comment by Marukon** [comment]
开发者说，In most case, set it to 0 can give you best result.

**Comment by Hymnu** [comment]
做了一个python小程序来实现对数词频转换：[Trans_Rime_to_Fcitx](https://github.com/Hymnu/Trans_Rime_to_Fcitx)
但要求输入是.tsv/.dict.yaml，稍作修改即可。
